### DISCUSSION TITLE
Device pairing

### DISCUSSION THREAD

**Discussion by rhelgeby** [strandborg/AlsavoCtrl#12] [General]
Excellent work on the reverse engineering.

Has anyone figured out how the initial device pairing works?

On the initial setup the heat pump is not connected to the Wi-Fi network, and is somehow receiving SSID and password. I'm wondering what the app is doing to send this. Maybe there's some special Wi-Fi broadcast packets it sends on the air? When I put the heat pump in setup mode maybe it's listening for such packets on the air. Not sure how to capture those packets yet. Maybe if Wireshark can listen listen on a Wi-Fi interface without being connected to any network.

I'm trying to avoid the Alsavo Pro app as much as possible. The privacy on the app and heat pump firmware is horrible. Turns out my heat pump also connect to 47.254.157.150 and send Wi-Fi SSID and password in cleartext. Combine that with the fact that the app insist on turning on GPS to work. Then the server owners have a nice dataset of Wi-Fi networks. Name, password and location. This can't be legal. Maybe in China, but not in Europe.

I have blocked internet access for the heat pump (it can only access the LAN subnet), and connections to 47.254.157.150 in my firewall. The app still seems to work on the local network, so that's ok. After forcing the app to work locally, it seems more robust as well.

**Comment by strandborg** [comment]
I am interested in this as well, just haven't had time or energy to dig into it yet. Wireshark etc captures are of course welcome (or better yet, ready-made pull requests =) )

**Reply by rhelgeby** [reply]
I'm not good in the C++ world. Instead I'm experimenting with a Java app myself since that's where I'm most efficient at the moment. Got past the handshake and next up is making queries. I will make a repository for it on my account soon. Just experimental stuff now.

Understanding the protocol is most important. Maybe I figure out something. It's a low priority hobby project here as well.

Like what these magic numbers in the auth packet are. Maybe they are specific to your device.
```
final int uuid1 = 0xffffffff;
final int uuid2 = 0xaceee2d3; // little endian: 0xd3e2eeac
final int uuid3 = 0;
final int uuid4 = 0x55c7fd6a; // little endian: 0x6afdc755
```

**Reply by strandborg** [reply]
Quite possible. I reversed the protocol by doing a few Wireshark captures plus then disassembling both java and native sides of the Alsavo Pro app as well as attaching a debugger into the app itself; the native lib of the app seems to support a wide variety of different IoT devices so only some parts of the code are relevant in our use case.

**Reply by rhelgeby** [reply]
It turns out macOS has a hidden gem called Wi-Fi Diagnostics. A nice app for technical analysis. One of the tools can scan a specific Wi-Fi channel and dump it to a pcap file. I can see everything, such as beacon frames, various probes and acks, and data packets (encrypted).

However, I bet the traffic from the heat pump is not encrypted.

**Reply by rhelgeby** [reply]
I think the pairing is done through Smart Config, depending on what kind of Wi-Fi device that's in the heat pump.

Maybe something similar to this: https://github.com/tuanpmt/iot-for-beginners-book/blob/master/doc/05-wifi-config/01-smartconfig.en.adoc

Note: It's still just speculation, but seems plausible.

According to the link above, the SSID and password are encoded in the length field of a broadcast packet the phone app is sending. One packet for each byte to transmit. The payload is random, but at the length corresponding to the byte being sent - if I understood this correctly. Even though the packet data is encrypted, the length field is not.

When I long press the standby button on the heat pump it goes in to setup mode. It's broadcasting probing packets every three seconds, until I exit this mode. Then it probably also listen for these encoded packets and decode the bytes. I don't know how it know which packets to read the length field from. Maybe the first bytes sent are some kind of signature so it knows this is a message with SSID and password. Not sure how this works in detail, but I suppose it's possible to capture these packets sent by the app.

It explains why the pairing is so unreliable and takes time. Last time it took multiple attempts before it connected.

**Reply by strandborg** [reply]
Yeah the pairing is super unreliable for me too, so I think you might be on the right track there.

**Reply by rhelgeby** [reply]
Found this: https://www.shielder.com/blog/2020/04/notsosmartconfig-broadcasting-wifi-credentials-over-the-air/

It refers to a proof of concept script that reads a pcap file and extracts the Wi-Fi credentials: https://github.com/ShielderSec/NotSoSmartConfig

It didn't find any credentials in my packet captures, though I was never able to complete pairing during a capture either.

The scripts also contains some comments on how the traffic is flowing. Useful info for later. I have other things to do, so I'll be on and off this project.

**Comment by rhelgeby** [comment]
I'm putting this aside for now. There should be enough info for somebody else to continue instead.

So far it looks like it's using the Smart Config protocol. In some implementations it also goes by ESP-Touch.

Here's a collection of things I found useful or interesting:

* ESP-Touch: https://www.espressif.com/sites/default/files/documentation/esp-touch_user_guide_en.pdf
* Technical document (see chapter 14, "Sniffer Introduction"), not necessarily for this device, but still helpful: https://www.espressif.com/sites/default/files/documentation/esp8266-technical_reference_en.pdf
* Another useful blog post about how Smart Config works: https://depletionregion.blogspot.com/2013/10/cc3000-smart-config-transmitting-ssid.html

On macOS I used the sniffer tool in the Wi-Fi Diagnostics app to capture packets during pairing. If you configure the Wi-Fi password in Wireshark you can also see the decrypted content, but that's not required. The heat pump only rely on the payload length while listening.

The app started with broadcasting packets with lengths that count 1, 2, and 3 repeatedly for a few seconds. This is for the heat pump device to detect which Wi-Fi channel to listen to. Then the app starts alternating between 1 byte payloads and various lengths for the encoded data. Then it goes back to counting and repeats the process, everything on fixed intervals.

I wasn't able to decode any of this because it's probably encrypted. The decompiled app has a hard coded AES key in it (in the same class that has the "AES/CBC/PKCS5Padding" string constant). I didn't go further with this, but joining all the nibbles to bytes and decrypting it with that key would probably reveal the Wi-Fi name and password.

At least it's doing something to avoid broadcasting the Wi-Fi password to all my neighbors. Though, either the heat pump or the app was still sending the password in cleartext to a remote server (47.254.157.150), so you should still block this server in the firewall. Also block internet access for the heat pump.

This AlsavoCtrl program and the heat pump still works with LAN-only access.

For now I'm mostly interested in collecting stats from the heat pump and working on a Java implementation for this.

**Comment by strandborg** [comment]
Thank you for your efforts!
